feat: create webrtc task

rodrigoleonello opened the pull request:
Depends on:
- [ ] https://github.com/tidewise/tidewise.common-package_set/pull/120
- [ ] https://github.com/tidewise/tools-rustysignal/pull/2

rodrigoleonello committed:
feat: create task configuration.

doudou commented on comms_webrtc.orogen:
```diff
@@ -2,54 +2,25 @@ name "comms_webrtc"
-task_context "Task" do
-    # This is the default from now on, and should not be removed. Rock will
-    # transition to a setup where all components use a configuration step.
+task_context "WebRTCTask" do
```
Just "Task" is a good name in a package that will probably have only one.

The webrtc part is already in the project name.

doudou commented on tasks/WebRTCTask.cpp:
```diff
@@ -0,0 +1,449 @@
+void WebRTCTask::onOffer(shared_ptr<rtc::WebSocket> wws) {
+    tmpPc.reset();
+
+    if (!this->getIdFromMessage(tmpPeerID)) {
```
Don't use `this->` when it is not needed.

doudou replied:
Not sure what you are trying to achieve with the `peerId` thing ?

doudou commented on tasks/WebRTCTask.cpp:
```diff
@@ -0,0 +1,449 @@
+
+    std::string getCandidate( ) {
+        auto const& field = jdata["candidate"];
+        return field.asString();
```
Pretty sure you can do `jdata["candidate"].asString();` (valid for the others)

doudou commented on tasks/WebRTCTask.cpp:
```diff
@@ -0,0 +1,449 @@
+    return true;
+}
+
+bool WebRTCTask::getIdFromMessage(std::string &out_str) {
```
Do not use out arguments (please have a read on our coding standards)

rodrigoleonello replied:
yeah, sure. When we are creating from scratch is easier to keeps the pattern. How I'm coding following the lib example, indeed there are a lot of changes to do.

doudou replied:
That's fair enough ... please tell me when I can start to talk about refactoring to be closer to our coding standard.

Tbh, I'm not entirely sure this is a very good strategy, but please follow your feelings in this respect.

doudou reviewed (commented):
So far so good ... only style stuff.

The one architectural question I have is: what's your plan for the signalling server ? The component assumes it can connect using websockets to a server, so where / what will this server be ?

rodrigoleonello committed:
fix: add cmake set c++ 17.

rodrigoleonello committed:
fix: remove unnecessary get methods, tmpPeerId and peerId variables

update the createPeerConnection to use Json msg instead nlohmann::Json,
modify the url created by the websocket according to
(wss://signalserverhost?user=yourname) and rename properties.

rodrigoleonello committed:
feat: modify the local_peer_id to be class variable.

rodrigoleonello committed:
fix: update websocket url.

rodrigoleonello committed:
fix: register missed callbacks.

rodrigoleonello committed:
feat: create input and output logics between the component and the data channel.

rodrigoleonello committed:
fix: check in the websocket callback if the peerconnection is already done and

turns the localpeeId a class member.

rodrigoleonello committed:
feat: create configurePeerConnection and configureDataChannel methods.

rodrigoleonello committed:
feat: create configureWebSocket method.

doudou commented on comms_webrtc.orogen:
```diff
@@ -2,54 +2,24 @@ name "comms_webrtc"
-    # A configuration property (here, a std::string). Its value can be retrieved
-    # in the C++ code with # _config_value.get() and _config_value.set(new_value).
-    property "config_value", "/std/string"
+    property "ice_server", "std/string"
+    property "local_peer_id", "std/string"
+    property "remote_peer_id", "std/string"
+    property "data_channel_label", "std/string"
+    property "websocket_server_name", "std/string"
```
Documentation please. Especially (but not only) what happens if `remote_peer_id` is empty.

doudou reviewed (commented):
Please remove the validation and check that things are present in the get method. Use exceptions for errors and return values instead of using in/out arguments

For instance

```
void validateFieldPresent(std::string fieldName) {
  if (!jdata.isMember(fieldName)) {
     throw MissingField("message does not contain the '" + fieldName + "' field");
  }
}

std::string getId() {
  validateFieldPresent("id");
  return jdata["id"].asString();
}

std::string Task::getIdFromMessage()
{
    return decoder->getId();
}
```

rodrigoleonello committed:
feat: create conversion to get binary data from the msg and write in the ouput.

doudou commented on tasks/Task.cpp:
```diff
@@ -310,7 +310,13 @@ void Task::configureDataChannel(
+                        std::vector<uint8_t> new_data;
+                        for(unsigned int i=0; i<data_byte.size(); i++)
+                        {
+                            new_data[i] = to_integer<uint8_t>(data_byte[i]);
```
This will crash, as you have not resized the array.

To avoid copying the vector, you could also use dataPacket.data directly.

```
dataPacket.data.resize(data_byte.size());
for(unsigned int i=0; i<data_byte.size(); i++) {
    dataPacket.data[i] = to_integer<uint8_t>(data_byte[i]);
}

rodrigoleonello committed:
fix: MessageDecoder struct methods refactoring.

rodrigoleonello committed:
feat: create datachannel send method to push raw data.

doudou commented on tasks/Task.cpp:
```diff
@@ -3,47 +3,355 @@
+                [&, wdc = make_weak_ptr(data_channel)]()
+                {
+                    if (auto data_channel = wdc.lock())
+                        data_channel->send("Hello from " + mLocalPeerId);
```
Pretty sure you mean to remove that ;-)

rodrigoleonello replied:
kkkkkk

doudou commented on tasks/Task.cpp:
```diff
@@ -3,47 +3,355 @@
+    RawPacket raw_packet;
+    if (_data_in.read(raw_packet) == RTT::NewData && !_remote_peer_id.get().empty())
+    {
+        shared_ptr<rtc::PeerConnection> peer_connection =
+            createPeerConnection(mWs, _remote_peer_id.get());
+
+        // We are the offerer, so create a data channel to initiate the process
+        string label = _data_channel_label.get();
+        LOG_INFO_S << "Creating DataChannel with label \"" << label << "\"" << std::endl;
+        shared_ptr<rtc::DataChannel> data_channel =
+            peer_connection->createDataChannel(label);
+
+        peer_connection->onDataChannel(
+            [&data_channel, raw_packet](shared_ptr<rtc::DataChannel> incoming)
+            {
+                data_channel = incoming;
+                vector<byte> data;
+                data.resize(raw_packet.data.size());
+                for(unsigned int i=0;i<raw_packet.data.size();i++)
+                {
+                    data[i] = static_cast<byte>(raw_packet.data[i]);
+                }
+                data_channel->send(&data.front(), data.size());
+            });
```
You are supposed to create a data channel once, and use it forever. Not create one per message !

rodrigoleonello replied:
> Start to look fine, but the signalling messages (in and out) do not match the format from rustysignal ...

yeah, I know. I'm still have doubt about that, so I decided to fix afterwards. I'm getting confusing about how to parser the msg.

doudou reviewed (commented):
Start to look fine, but the signalling messages (in and out) do not match the format from rustysignal ...

rodrigoleonello committed:
feat: update i/o and properties comments.

rodrigoleonello committed:
feat: update method name to clarify the undestanding.

doudou commented on tasks/Task.cpp:
```diff
@@ -3,47 +3,356 @@
 }
 
-Task::~Task()
+bool Task::parseIncomingMessage(char const* data)
```
Use  exceptions to handle errors.

doudou commented on tasks/Task.cpp:
```diff
@@ -3,47 +3,356 @@
+        [&, wws](rtc::Description description)
+        {
+            Json::Value message;
+            Json::FastWriter fast;
```
Define variables as close as possible to their first use.

rodrigoleonello committed:
fix: register the datachannel callback once instead to always that a newData is received

and update the configure websocket step to check if already have a peerconnection
before create one, if the type received by the msg is "offer". Moreover, remove the
map variables since it is possible to check if the variable is null (otherwise, it was started once)

rodrigoleonello committed:
fix: starts the datachannel on the startHook and separate peerConnection

into two others methods - initiatepeerconnection and createPeerConnectionOnOffer

rodrigoleonello committed:
feat: modify the websocket msg type according to the msg type used in the signaling server.

rodrigoleonello committed:
feat: move the MessageDecoder struct to comms_webrtcTypes.

rodrigoleonello committed:
feat: create datachannel msg string to vector<uint8_t> conversion.

rodrigoleonello committed:
fix: initialize the websocket ptr and remove the MessageDecoder ptr.

rodrigoleonello committed:
fix: configure datachannel in the startHook for the case that we have a remote peer id.

add protocol as argument in the initiatePeerConnection to differentiate either when we are receiving the msg in the callback or creating one by the property

rodrigoleonello committed:
feat: update property comments and enable test.

rodrigoleonello committed:
fix: create peer conecction in onAnswer and OnCandidate methots

Simplify the json msg structure.

rodrigoleonello committed:
feat: add test file.

rodrigoleonello committed:
feat: create status output.

doudou commented on comms_webrtc.orogen:
```diff
@@ -2,54 +2,48 @@ name "comms_webrtc"
+import_types_from "base"
+
+using_library "libdatachannel"
+using_library "jsoncpp", typekit: true
```
`typekit: true` is the default. Both libraries could be imported with `typekit: false` though...

rodrigoleonello replied:
If I put `false`, it is not possible to include `"json/json.h"`in the `comms_webrtcTypes.hpp`. 
It says that can't find the header..

doudou replied:
That would be because MessageDecoder is there. And should not be.

Only put types there that are to be used on the component(s) interface(s)

You can put the rest in separate headers, for instance in e.g. `tasks/MessageDecoder.hpp`

doudou commented on comms_webrtc.orogen:
```diff
@@ -2,54 +2,48 @@ name "comms_webrtc"
+    property "data_channel_label", "std/string"
+
+    # Websocket server name (e.g. 127.0.0.1:3012)
+    property "websocket_server_name", "std/string"
```
Stick to the webrtc naming. This is the signalling server.

doudou commented on manifest.xml:
```diff
@@ -1,17 +1,15 @@
+  <test_depend package="tools/syskit" />
+  <test_depend package="tools/libdatachannel" />
+  <test_depend package="drivers/orogen/iodrivers_base" />
+  <test_depend package="jsoncpp" />
```
`test_depend` is needed only for packages that are not already build dependencies.

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,331 @@
+            message["description"] = string(description);
+
+            if (auto ws = make_weak_ptr(mWs).lock())
+                ws->send(fast.write(message));
```
Style ... always use brackets.

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,331 @@
+        [&, protocol, remote_peer_id](rtc::Candidate candidate)
+        {
+            Json::Value message;
+            Json::FastWriter fast;
```
Same

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,331 @@
+            message["candidate"] = string(candidate);
+            message["mid"] = candidate.mid();
+
+            if (auto ws = make_weak_ptr(mWs).lock())
```
Same

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,331 @@
+            Json::FastWriter fast;
+            message["protocol"] = protocol;
+            message["to"] = remote_peer_id;
+            message["actiontype"] = "candidate";
```
Per the rustysignal documentation, it is `action`, not `actiontype`

rodrigoleonello replied:
When I was trying to change msg through the websocket I tried to fill the structure with different fields to test what happens, and only for the `protocol` and `to` fields that I had issues. But in any case, you are right...let's keep what the documentation says.

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,331 @@
+
+void Task::createPeerConnection()
+{
+    string protocol = mDecoder.getProtocol();
```
I really don't think this makes a lot of sense. Hardcode `one-to-one` for now.

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,331 @@
+                       << std::endl;
+
+            mDataChannel = data_channel;
+            data_channel->onOpen([&, wdc = make_weak_ptr(data_channel)]() {});
```
I think you should be filling something in the status structure when this happens ...

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,331 @@
+
+    mWs->onOpen(
+        [&ws_promise]()
+        {
+            LOG_INFO_S << "WebSocket connected, signaling ready" << std::endl;
+            ws_promise.set_value();
```
Same ... should have some status field set when the websocket is opened and when it closes.

doudou reviewed (commented):
You are rather obviously not following rustysignal documentation. It may work, but that would be luck.

Stick to `{ protocol: "one-to-one", to: "receiver_username", "action": actiontype, "data": data };`

Splitting the signalling from the data would also make things clearer.

rodrigoleonello committed:
fix: refactor the signaling msg structure and throw in case the parser fails.

Remove unnecessary test_depends and create the MessageDecoder structure in a separate header.

rodrigoleonello committed:
feat: update component status struct.

rodrigoleonello committed:
feat: modify datachannel condition check before sending msg.

rodrigoleonello committed:
fix: change protocol name.

doudou commented on comms_webrtcTypes.hpp:
```diff
@@ -1,15 +1,74 @@
+        HaveRemotePranswer
+    };
+
+    enum LocalDescription
```
I think it would be best to suffix all these with `State` (all enums / structs in the file)

doudou commented on tasks/MessageDecoder.hpp:
```diff
@@ -0,0 +1,74 @@
+#define COMMS_WEBRTC_MESSAGE_DECODER_HPP
+
+#include "json/json.h"
+#include "stdio.h"
+#include "string.h"
+#include "memory"
```
I don't see anything in there that would require these headers... Sure about their need ?

rodrigoleonello replied:
indeed, it is not necessary.

doudou commented on tasks/MessageDecoder.hpp:
```diff
@@ -0,0 +1,74 @@
+            return reader->parse(data, data + strlen(data), &jdata, &errors);
+        }
+
+        void validateFieldPresent(std::string const &fieldName)
+        {
+            if (!jdata.isMember(fieldName))
+            {
+                throw std::invalid_argument(
+                    "message does not contain the " + fieldName + " field");
+            }
+        }
+
+        void validateDataFieldPresent(std::string const &fieldName)
```
*This* is why you avoid passing state through an attribute.

Just change `validateFieldPresent` to have a `Json::Value&` argument as first argument and boom ... you only need one method.

rodrigoleonello replied:
nice

doudou commented on tasks/MessageDecoder.hpp:
```diff
@@ -0,0 +1,74 @@
+
+        std::string getDescription()
+        {
+            validateDataFieldPresent("description");
```
You should check that `data` is present first.

doudou commented on test/webrtc_test.rb:
```diff
@@ -0,0 +1,48 @@
+        data_in.time = Time.now
+        data_in.data = [1, 0, 1, 0, 1, 1, 1, 0]
+
+        task1_output = expect_execution do
+            syskit_write task1.data_in_port, data_in
+        end.to do
+            have_one_new_sample(task1.status_port)
+        end
```
You are not waiting for the data channel to be established before you write ...

doudou commented on test/webrtc_test.rb:
```diff
@@ -0,0 +1,48 @@
+        end.to do
+            have_one_new_sample(task1.status_port)
+        end
+    end
```
And you must check that the received sample matches what was sent

Also, you need to check that task2 can talk to task1. Since connection establishment is assymetric, it might not work.

rodrigoleonello committed:
feat: modify enum/struct names and simplify validadeFildPresent method.

rodrigoleonello committed:
fix: create getFrom and fix the msg created in the answer case ...

Remove createPeerConnection from onAnswer and onCandidate since at this time for sure
we already have peer connection.

Create ping/pong methods to check if the remote id is reachable before starts to sending msg. Update tests.

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,422 @@
+        while(!mRemotePeerAnswerReceived)
+        {
+            ping();
+            usleep(100);
```
Waaaaay too quick. Should be 100ms (usleep is in microseconds).

Also, use the C++ sleep API

rodrigoleonello replied:
ok

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,422 @@
+        {
+            Json::Value message;
+            message["protocol"] = "one-to-one";
+            message["to"] = remote_peer_id;
```
~~I'm a bit skeptical about this ... I think that the `remote_peer_id` is specific to webrtc, and not to our signalling protocol. Use the `from` field here too.~~

Generally speaking, I think you really need to create a sendSignallingMessage(message) that would fill the protocol, to and from field ... to avoid forgetting something.

doudou replied:
And instead of referring to mDecoder, set a mRemoteID field.

- either using the remote peer id property in case it's not empty
- or when receiving the offer message

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,422 @@
+            {
+                Json::FastWriter fast;
+                ws->send(fast.write(message));
+                mState.peer_connection.local_description = NewDescription;
```
This handling of mState's local_description is misleading. The false branch is not that there was no local description, it is that you did not send it.

Tbh, I think it's irrelevant here. If the websocket is being destroyed, it means we're shutting down the component and it's fine to not send anything.

(This applies to the other flags dependent on write/no-write)

rodrigoleonello committed:
fix: remove misleading states and grow up the sleep time.

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,413 @@
-Task::~Task()
+void Task::configurePeerDataChannel(string const &remote_peer_id)
 {
+    mPeerConnection->onDataChannel(
```
Still wrong. The issue is that `onDataChannel` is called only on the "passive" peer. The active peer (the one that creates the data channel) will not get the onDataChannel event

However, in the 'active peer' case, you do have to set up the callbacks on the data channel object itself.

rodrigoleonello committed:
fix: move the datachannel callback register to a separate method and ...

remove that from the offer workflow.

doudou committed:
fix: use a mRemotePeerID attribute to store the ID of the remote peer

In both cases (active and passive).

doudou committed:
fix: display incoming data channel in onDataChannel

Accessing DataChannel::label() in onOpen makes the process crash. Likes
this, tests pass and valgrind does not complain

doudou committed:
fix: wait for the data channel to be established in the passive case as well

doudou committed:
fix: check that we do transmit data in the test

rodrigoleonello committed:
fix: remove print to debug and add test depend.

rodrigoleonello committed:
fix: update default values for constructor.

rodrigoleonello committed:
feat: close both datachannel and peerconection on stopHook.

The onClose websocket is already called after calls the peerconnection close.

rodrigoleonello committed:
feat: update test

doudou commented on comms_webrtcTypes.hpp:
```diff
@@ -54,7 +56,7 @@ namespace comms_webrtc
     struct WebRTCState
     {
-        PeerConnectionState peer_connection;
+        PeerConnectionState peer_connection = PeerConnectionState();
```
Structures are always initialized by their default constructor.

doudou commented on comms_webrtcTypes.hpp:
```diff
@@ -37,6 +37,8 @@ namespace comms_webrtc
         GatheringState gathering_state = NoGathering;
         SignalingState signaling_state = NoSignaling;
+
+        PeerConnectionState() {}
```
I am assuming that you did this because of a compiler error

If not, remove it.

If yes, a better way to do this in C++11 is `PeerConnectionState() = default;` (cf. https://en.cppreference.com/w/cpp/language/default_constructor)

rodrigoleonello replied:
nice!

doudou commented on tasks/Task.cpp:
```diff
@@ -414,5 +417,23 @@ void Task::updateHook()
+    {
+        std::future<void> pc_close_future = mPeerConnectionClosePromise.get_future();
+        mPeerConnection->close();
+        pc_close_future.get();
```
The if is against the pointer, but you don't reset the pointer.

Either test against isOpen, or reset the shared pointer. (I think the latter is better)

doudou commented on test/webrtc_test.rb:
```diff
@@ -31,7 +31,7 @@
     end
 
-    it "configure and starts the tasks" do
+    it "configure, starts and send message between the tasks" do
```
Don't attempt to cram every test in a single `it`

Instead, test one thing after the other in different `it` calls, optionally using helpers to avoid too much repetition

Tests that I see

```
it "emits start only after the data channel is fully established" # check the status port
it "transmits data from the active to the passive task once started"
it "transmits data from the passive to the active task once started"
it "detects that the channel has been closed from the passive side" # task2 should get into exception if task1 is stopped
it "detects that the channel has been closed from the active side" # task1 should get into exception if task1 is stopped
it "successfully reestablishes connection after a stop/start cycle" # to make sure the component is functional if we stop and then start it again
```

rodrigoleonello replied:
Do you really think that we should go into an exception event if the datachannel is closed? Because nowadays the component is not going to that kind of event in that scenario.

doudou replied:
I really think so, yes. It's actually pretty critical.

> Because nowadays the component is not going to that kind of event in that scenario.

What do you mean ? That none of the callbacks fire ? Or that the component itself does not handle it ?

rodrigoleonello replied:
The fact that we go into the onClosed callback doesn't represent a exception
Should we put an exception to represent that on the callback?

rodrigoleonello replied:
Should we put an exception to represent that on the callback?

rodrigoleonello committed:
fix: split test into different scopes

rodrigoleonello committed:
feat: remove unnecessary struct default constructor.

rodrigoleonello committed:
fix: reset the peerconnection ptr in the stopHook.

rodrigoleonello committed:
fix: remove check redundancy

rodrigoleonello committed:
feat: organize the promises to avoid get stuck for the active and passive cases

Moreover, create timeout to the remote peer check in the active case, and a timeout
related to the datachannel open, in the passive case. Update the tests.

rodrigoleonello committed:
feat: update enum names

rodrigoleonello committed:
feat: use state as condition in the updatehook and throw if the datachannel is closed

rodrigoleonello committed:
feat: create evaluate method to check the datachannel status.

rodrigoleonello committed:
feat: create evaluate method to check websocket states.

doudou commented on test/webrtc_test.rb:
```diff
@@ -0,0 +1,189 @@
+        deploy("task_a", "task_b", Time.at(5))
+        configure_and_start
+
+        data_in_task1 = Types.iodrivers_base.RawPacket.new
+        data_in_task1.time = Time.now
+        data_in_task1.data = [1, 0, 1, 0, 1, 1, 1, 0]
+
+        expect_execution do
+            syskit_write task1.data_in_port, data_in_task1
+        end.to { have_one_new_sample task2.data_out_port }
+
```
That should not be here. We are testing that task2 is able to send data to task1. I don't expect we have to send data from task1 to task2 first, or ?

rodrigoleonello replied:
Removing...you are right.

doudou commented on test/webrtc_test.rb:
```diff
@@ -0,0 +1,189 @@
+
+        deploy("task_c", "task_d", Time.at(5))
+        configure_and_start
+    end
```
you do not test that the new tasks are functional.

rodrigoleonello replied:
ooh, sorry. Fixing that..

doudou commented on test/webrtc_test.rb:
```diff
@@ -0,0 +1,189 @@
+    end
+
+    it "returns exception when the wait to find the remote peer id exceeds the timeout" do
+        deploy("task_a", "task_b", Time.at(0))
+        syskit_configure(task1)
+        syskit_configure(task2)
+
+        expect_execution do
+            task1.start!
+            task2.start!
+        end.to do
+            fail_to_start task1
+            fail_to_start task2
+        end
```
When possible, use real-world cases in your tests. A zero time out is not.

Just start a single task with a non-zero timeout. It should timeout because *the remote task is not there*, not because of some weird configuration values.

doudou commented on test/webrtc_test.rb:
```diff
@@ -0,0 +1,189 @@
+        deploy("task_a", "task_b", Time.at(2))
+        configure_and_start
+
+        data_in = Types.iodrivers_base.RawPacket.new
+        data_in.time = Time.now
+        data_in.data = [1, 0, 1, 0, 1, 1, 1, 0]
+
+        data = expect_execution do
+            syskit_write task1.data_in_port, data_in
+        end.to { have_one_new_sample task2.data_out_port }
+
+        assert_equal data_in.data, data.data
```
Not needed.

doudou commented on test/webrtc_test.rb:
```diff
@@ -0,0 +1,189 @@
+        deploy("task_a", "task_b", Time.at(2))
+        configure_and_start
+
+        data_in = Types.iodrivers_base.RawPacket.new
+        data_in.time = Time.now
+        data_in.data = [1, 0, 1, 0, 1, 1, 1, 0]
+
+        data = expect_execution do
+            syskit_write task1.data_in_port, data_in
+        end.to { have_one_new_sample task2.data_out_port }
```
Not needed.

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,496 @@
+    // wss://signalserverhost?user=yourname
+    const string url = _signaling_server_name.get() + "?user=" + _local_peer_id.get();
+    mWs->open(url);
+    ws_future.get();
```
You should have a timeout for this as well.

Actually, you should have a timeout for *every* .get() you call on a future

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,496 @@
+    {
+        std::future<void> dc_future = mDataChannelPromise.get_future();
+        base::Time start_time = base::Time::now();
+        // Wait for the datachannel open until the timeout
```
There is a way to wait for a future "until a certain time", we saw that together in one of our calls. Use this, which will avoid the need to have two different branches.

doudou replied:
https://cplusplus.com/reference/future/future/wait_until/

rodrigoleonello replied:
I got confuse with the other one (wait_for)...which is no good for our use case. That other one you mentioned fits, thanks!

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,496 @@
+        mRemotePeerID = _remote_peer_id.get();
+        mRemotePeerAnswerReceived = false;
+
+        base::Time start_time = base::Time::now();
+        // Get contact with the remote peer and create datachannel
+        while (!mRemotePeerAnswerReceived)
+        {
+            base::Time duration = base::Time::now() - start_time;
+            if (duration > _wait_remote_peer_time_out.get())
+            {
+                mWaitRemotePeerPromise.set_exception(
+                    std::make_exception_ptr(std::runtime_error("Timeout to find the remote peer")));
+                break;
+            }
+            ping();
+            sleep_for(100ms);
+        }
```
There is no future in this case, just the ping and mRemotePeerAnswerReceived

doudou replied:
And I'd suggest moving the whole block of code into a 'waitRemotePeerPresent' method or something.

rodrigoleonello replied:
> There is no future in this case, just the ping and mRemotePeerAnswerReceived

I didn't understand....and I'm still thinking how to use `wait_for` or `wait_until`methods in that situation, because will block the thread (so it is not possible to use inside the loop, and don't make sense to use outside because if we are at this point, we already have a `set_value`)

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,496 @@
+void Task::errorHook() { TaskBase::errorHook(); }
 void Task::stopHook()
 {
+    if (mDataChannel->isOpen())
+    {
+        std::future<void> dc_close_future = mDataChannelClosePromise.get_future();
+        mDataChannel->close();
+        dc_close_future.get();
+    }
+    if (mPeerConnection)
```
I would have expected you to follow the same pattern in both ifs ... either open/close or delete.

rodrigoleonello replied:
There is not isOpen method for the peerConnection...I will just check the pointer in both cases.

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,496 @@
+        string label = _data_channel_label.get();
+        mDataChannel = mPeerConnection->createDataChannel(label);
+        registerDataChannelCallBacks(mDataChannel);
+        dc_future.get();
```
You also need a timeout here. And if it does time out, you need to delete the peer connection and the data channel

rodrigoleonello replied:
I will do the same (delete) for the websocket pointer, ok?

doudou commented on test/webrtc_test.rb:
```diff
@@ -0,0 +1,189 @@
+        configure_and_start
+    end
+
+    it "returns exception when the wait to find the remote peer id exceeds the timeout" do
```
You can also add a test for the websocket connection timeout by passing 240.24.15.234 as the signalling server IP, and trying to configure it (see https://www.rock-robotics.org/rock-and-syskit/components/testing.html#testing-configuration-or-start-failures for that last bit)

doudou commented on test/webrtc_test.rb:
```diff
@@ -0,0 +1,189 @@
+    run_live
+
+    attr_reader :task1, :task2
+
```
One other thing: you will have to start rustysignal in the `before` and stop it in `after`. You can't expect it to be already running during the tests.

rodrigoleonello replied:
Do you remember where could I find an example to do that?

rodrigoleonello committed:
feat: create timeout for websocket, datachannel and peerconnection

During the start and stop is needed to check if the procedure is
feasibly until the timeout, otherwise an exception event is emitted.

rodrigoleonello committed:
feat: add websocket timeout exception test.

rodrigoleonello committed:
feat: spawn and kill automatically the signaling server process in the test.

rodrigoleonello committed:
feat: use wait_until instead the while loop to wait for the remote peer answer.

doudou commented on test/webrtc_test.rb:
```diff
@@ -0,0 +1,202 @@
+    before do
+        time = Time.now
+        timeout = Time.now + 1
+        while time < timeout
```
Just `while Time.now < timeout` would do

doudou commented on test/webrtc_test.rb:
```diff
@@ -0,0 +1,202 @@
+
+    attr_reader :task1, :task2
+
+    run_pid = nil
```
Even if it does work, this is bad practice. A describe() block essentially defines a class, use instance variables (@rustysignal_pid)

doudou commented on test/webrtc_test.rb:
```diff
@@ -0,0 +1,202 @@
+                sleep(0.01)
+            end
+            time = Time.now
+        end
```
You do not fail on timeout

doudou commented on test/webrtc_test.rb:
```diff
@@ -0,0 +1,202 @@
+    after do
+        unless run_pid.nil?
+            Process.kill("KILL", run_pid)
+            Process.wait
```
`Process.waitpid @rustysignal_pid`

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,525 @@
+    const string url = _signaling_server_name.get() + "?user=" + _local_peer_id.get();
+    mWs->open(url);
+    future_status status =
+        ws_future.wait_until(chrono::system_clock::now() +
```
Why don't you use `wait_for` ? Which is ... defined for this very purpose ?

rodrigoleonello replied:
I don't know too kkk
Changing...

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,525 @@
+    mWs->open(url);
+    future_status status =
+        ws_future.wait_until(chrono::system_clock::now() +
+                             chrono::seconds(static_cast<int>(_websocket_open_time_out.get().toSeconds())));
```
`_websocket_open_time_out` is a `base::Time`. It has a microsecond resolution. Converting to seconds like this makes sure that one day someone will pick a sub-second value and things will break. Use microseconds.

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,525 @@
+    else
+    {
+        mWs.reset();
+        throw runtime_error("WebSocket open failed");
```
"Timed out waiting for the websocket connection to be ready"

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,525 @@
+        mRemotePeerAnswerReceived = false;
+        future<void> wait_remote_peer_future = mWaitRemotePeerPromise.get_future();
+        // Try to get contact with the remote peer and create datachannel
+        ping();
+        future_status status =
+            wait_remote_peer_future.wait_until(chrono::system_clock::now() +
+                                               chrono::seconds(static_cast<int>(_wait_remote_peer_time_out.get().toSeconds())));
```
Here, the while loop does make sense. It ensures that if we are configuring/starting two components at roughly the same time, they have a chance to find each other.

doudou replied:
Also, same comments about seconds and wait_for

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,525 @@
+    if (!_remote_peer_id.get().empty())
+    {
+        mRemotePeerID = _remote_peer_id.get();
+        mRemotePeerAnswerReceived = false;
```
You don't need that flag anymore

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,525 @@
+        }
+        else
+        {
+            throw runtime_error("DataChannel close failed");
```
"Timeout waiting for the data channel to close"

doudou commented on tasks/Task.cpp:
```diff
@@ -2,48 +2,525 @@
+        }
+        else
+        {
+            throw runtime_error("Peer connection close failed");
```
"Timeout waiting for the peer connection to close"

doudou commented on test/webrtc_test.rb:
```diff
@@ -0,0 +1,202 @@
+
+    it "emits start only after the data channel is fully established" do
+        deploy("task_a", "task_b", Time.at(2), "127.0.0.1:3012")
+        configure_and_start
```
You do have to test something. Like for instance value on the status port ?

I would suggest that you write a status sample at the end of the startHook and read and return it from `configure_and_start`.

doudou commented on test/webrtc_test.rb:
```diff
@@ -0,0 +1,202 @@
+        deploy("task_a", "task_b", Time.at(5), "127.0.0.1:3012")
+        configure_and_start
+
+        data_in = Types.iodrivers_base.RawPacket.new
+        data_in.time = Time.now
+        data_in.data = [1, 0, 1, 0, 1, 1, 1, 0]
+
+        data = expect_execution do
+            syskit_write task1.data_in_port, data_in
+        end.to { have_one_new_sample task2.data_out_port }
+
+        assert_equal data_in.data, data.data
```
You've already tested for this. No need to repeat it here.

doudou commented on test/webrtc_test.rb:
```diff
@@ -0,0 +1,202 @@
+
+        assert_equal data_in.data, data.data
+
+        state_task = Types.comms_webrtc.WebRTCState.new
+        state_task.peer_connection.state = :NoConnection
+        state_task.peer_connection.gathering_state = :NoGathering
+        state_task.peer_connection.signaling_state = :NoSignaling
+        state_task.data_channel = :NoDataChannel
+        state_task.web_socket = :NoWebSocket
+
+        state = expect_execution do
+            task1.stop!
+            task2.stop!
+        end.to do
+            emit task1.stop_event
+            emit task2.stop_event
+            [have_one_new_sample(task1.status_port),
+             have_one_new_sample(task2.status_port)]
+        end
+
+        assert_equal state_task, state[0]
+        assert_equal state_task, state[1]
```
This should be its own `it "disconnects the WebRTC connection on stop"` test.

doudou commented on test/webrtc_test.rb:
```diff
@@ -0,0 +1,202 @@
+        assert_equal data_in.data, output.data
+    end
+
+    it "returns exception when the wait to find the remote peer id exceeds the timeout" do
```
I'm afraid you have to test that from both sides, since their codepaths are different.

rodrigoleonello replied:
I will do it.

doudou commented on test/webrtc_test.rb:
```diff
@@ -0,0 +1,202 @@
+        end
+    end
+
+    it "returns timeout exception when the websocket is not initialized" do
```
`... when the websocket connection times out`

doudou reviewed (commented):
Almost there ...

rodrigoleonello committed:
feat: change wait_until method by wait_for and update the while loop logic to check if the remote peer is reachable.

rodrigoleonello committed:
feat: put status output on the start and update default values for the peer_connection callbacks

rodrigoleonello committed:
fix: update tests to check the status output and fix the while loop logic for spawn the rustysignal process

doudou commented on tasks/Task.cpp:
```diff
@@ -434,40 +433,45 @@ bool Task::startHook()
         {
-            throw runtime_error("Timeout to get contact with the remote peer");
+            ping();
+            if (wait_remote_peer_future.wait_for(
+                    chrono::microseconds(_wait_remote_peer_time_out.get().toMicroseconds())) == future_status::ready)
+            {
+                wait_remote_peer_future.get();
+                break;
+            }
+            else if (wait_remote_peer_future.wait_for(
+                    chrono::microseconds(_wait_remote_peer_time_out.get().toMicroseconds())) == future_status::timeout)
```
This does not solve the issue of what happens if we start the two components at the same time. When this happens, it might be that the first ping is lost. This is why in my example code I had a timeout for the wait_for (which should be the ping() period) and a timeout to break out of the loop (which is the actual _wait_remote_peer_time_out)


Moreover, you call wait_for twice, where you should save the return value and test the different possible values.
This is always going wait twice. Plus, the goal of the while loop is to send the ping more than once.

rodrigoleonello replied:
Now I got it. 
It's confuse...thanks!

doudou commented on test/webrtc_test.rb:
```diff
@@ -113,15 +123,26 @@ def configure_and_start()
-        data = expect_execution do
-            syskit_write task1.data_in_port, data_in
-        end.to { have_one_new_sample task2.data_out_port }
+        assert_equal :NoDataChannel, output[0].data_channel
+        assert_equal :NoDataChannel, output[1].data_channel
```
You are already testing that stops closes the data channel in another test ... no need to repeat it here.

These are *unit tests*. Keep things clear by testing what you mean to test (which here is it "successfully reestablishes connection after a stop/start cycle") and not more.

doudou commented on test/webrtc_test.rb:
```diff
@@ -142,53 +163,71 @@ def configure_and_start()
+            have_one_new_sample(task1.status_port)
+        end
+
+        assert_equal :NoDataChannel, output.data_channel
```
:+1:

doudou commented on test/webrtc_test.rb:
```diff
@@ -142,53 +163,71 @@ def configure_and_start()
 
         assert_equal state_task, state[0]
         assert_equal state_task, state[1]
```
The websocket should be valid after a stop. Only a cleanup should kill it. (`state_task.web_socket = :NoWebSocket` above)

rodrigoleonello replied:
Indeed...I will work on that.
Always after close datachannel and peerconnection the websocket OnClosed callback is called, so I will put a promise on the stopHook, as well, to properly wait for that close, and in the cleanup I will reset the pointer.

rodrigoleonello committed:
fix: properly close websocket in the stopHook and reset in the cleanup.

doudou commented on test/webrtc_test.rb:
```diff
@@ -0,0 +1,236 @@
+    attr_reader :task1, :task2, :rustysignal_pid
+
+    before do
+        timeout = Time.now + 1
+        while Time.now < timeout
+            begin
+                @rustysignal_pid = Process.spawn("rustysignal", "127.0.0.1:3012")
+                break
+            rescue Errno::ECONNREFUSED
+                sleep(0.01)
+            end
+        end
+        if @rustysignal_pid.nil?
+            raise "Timeout to start rustysignal"
+        end
```
This does not do anything of the stuff it's supposed to do.

- what does Process.spawn does ?
- why would you get Errno::ECONNREFUSED ?
- why would you get a nil @rustysignal_pid ?

rodrigoleonello replied:
The nil check actually was tied by a misunderstanding and the wrong use of the logic...

rodrigoleonello committed:
fix: update the while logic to check if the socket is open in the while after spawn the process
